wit/store.wit

```diff
@@ -0,0 +1,120 @@
+    /// depending on the specific implementation. For example:
+    ///
+    /// 1. Amazon DynamoDB calls a collection of key-value pairs a table
+    /// 2. Redis has hashes, sets, and sorted sets as different types of collections
+    /// 3. Cassandra calls a collection of key-value pairs a column family
+    /// 4. MongoDB calls a collection of key-value pairs a collection
+    /// 5. Riak calls a collection of key-value pairs a bucket
+    /// 6. Memcached calls a collection of key-value pairs a slab
+    /// 7. Azure Cosmos DB calls a collection of key-value pairs a container
+    ///
+    /// In this interface, we use the term `bucket` to refer to a collection of key-value pairs
+    resource bucket {
+        /// Get the store with the specified identifier.
+        ///
+        /// `identifier` must refer to a bucket provided by the host.
+        ///
+        /// `error::no-such-store` will be raised if the `identifier` is not recognized.
+        open: func(identifier: string) -> result<bucket, error>;
+
+        /// Get the value associated with the specified `key`
+        ///
+        /// The value is returned as an option. If the key-value pair exists in the
+        /// store, it returns `Ok(value)`. If the key does not exist in the
+        /// store, it returns `Ok(none)`. 
+        ///
+        /// If any other error occurs, it returns an `Err(error)`.
+        get: func(key: string) -> result<option<list<u8>>, error>;
+
+        /// Set the value associated with the key in the store. If the key already
+        /// exists in the store, it overwrites the value.
+        ///
+        /// If the key does not exist in the store, it creates a new key-value pair.
+        /// 
+        /// If any other error occurs, it returns an `Err(error)`.
+        set: func(key: string, value: list<u8>) -> result<_, error>;
+
+        /// Delete the key-value pair associated with the key in the store.
+        /// 
+        /// If the key does not exist in the store, it does nothing.
+        ///
+        /// If any other error occurs, it returns an `Err(error)`.
+        delete: func(key: string) -> result<_, error>;
+
+        /// Check if the key exists in the store.
+        /// 
+        /// If the key exists in the store, it returns `Ok(true)`. If the key does
+        /// not exist in the store, it returns `Ok(false)`.
+        /// 
+        /// If any other error occurs, it returns an `Err(error)`.
+        exists: func(key: string) -> result<bool, error>;
+
+        /// Get all the keys in the store. It returns a list of keys. Please note that for most
+        /// KeyValue implementations, this is a can be a very expensive operation and so it should
+        /// be used judiciously. Implementations can return any number of keys in a single response,
+        /// but MAY NOT return more than 2 MB of data in a single response. Any response should
```

Comment:
@lukewagner @Mossaka I kinda just threw out a number here for the max limit of data. Is there a more logical number here? I figure we need to provide some sort of direction of the maximum size any implementation should return in one chunk as opposed to a certain count of keys

Reply:
Good question.  I noticed that [wasi-io's streams.wit](https://github.com/WebAssembly/wasi-io/blob/main/wit/streams.wit), which has the same general question (how big can the `list<u8>` be) doesn't specify any particular limit.  If we want to avoid a magic constant, perhaps the text can specify the desired outcome: hosts must use their knowledge of the resource constraints of the system to pick a good chunk size that avoids OOM or excessive memory usage.

Reply:
Ok, so I can add a hint for max size. Let me do that and see what people think

Reply:
@kate-goldenring @lukewagner @Mossaka ok last ping I promise. I added a max-size hint so the caller can tell the implementation its max size. Does this API look ok or is it terrible?

Reply:
Perhaps `max-size` makes sense as a power-user option, but my impression is that most folks won't want to think about it and would rather the host just "do something reasonable".  I don't know if my previous comment sounded like it was suggesting adding a `max-size` parameter in which case, sorry about that, I wasn't; I was just saying that, instead of fixing 2MB in the doc comments, the doc comments could instead say that it's the host's job to pick an appropriate `max-size`.  Thus, I'd suggest either removing `max-size` or making it `option<u64>`.

Reply:
Ok, I'll remove it to keep it simple

Reply:
Reverted and added documentation for that